sql: fix placeholder type resolution

mgartner opened the pull request:
#### sql: fix placeholder type resolution

When assignment casts for inserts were add in #70722, type resolution of
placeholders in prepared statements was changed so that during the
prepare phase the optimizer would only plan assignment casts when it
could not be guaranteed that an insert column would be an identical type
to the target column. Unfortunately, the change to placeholder type
resolution change strays from Postgres's behavior and breaks some
third-party tools. This commit reverts the type resolution changes, and
as a result, assignment casts are always planned to convert each insert
column to its target column type, even if the cast ends up being a no-op
because the types are identical.

I attempted to implement a solution that did not plan assignment casts
in all cases, but it proved to be tedious and error prone, so I opted
for this simpler solution for now. My attempt required annotating each
placeholder expression in the memo with a context: a non-assignment
context or an assignment context. This context would be used during
the EXECUTE to determine whether a placeholder value should undergo a
regular cast or an assignment cast: a critical decision due to their
slightly different behaviors (see `PerformAssignmentCast`). The
problem is that correctly annotating placeholders in complex mutations
with expressions like subqueries or CTEs is difficult.

Fixes #71576

There is no release note because this bug was never present in a
release.

Release note: None

#### opt: add assignment cast normalization rules

This commit adds three new normalization rules related to assignment
casts. These rules work together to eliminate and fold assignment casts
and projections when possible.

These rules are important for allowing fast-path inserts to be planned
in more cases. Fast-path inserts can only be planned when the insert
input is a Values expression, so eliminating assignment cast projections
in mutation inputs is required for this optimization.

These rules also help eliminate uniqueness checks for
`gen_random_uuid()` values in more cases. By pushing down and
eliminating assignment casts that wrap `gen_random_uuid()`, the
optimizer will not plan uniqueness checks in some cases because it can
recognize the builtin function expression.

Release note: None

#### opt: add assign-placeholders-build opt test directive

This commit adds the `assign-placeholders-build` opt test directive. It
behaves similar to `assign-placeholders-norm`, but disables
normalization rules.

Release note: None

mgartner committed:
sql: fix placeholder type resolution

When assignment casts for inserts were add in #70722, type resolution of
placeholders in prepared statements was changed so that during the
prepare phase the optimizer would only plan assignment casts when it
could not be guaranteed that an insert column would be an identical type
to the target column. Unfortunately, the change to placeholder type
resolution change strays from Postgres's behavior and breaks some
third-party tools. This commit reverts the type resolution changes, and
as a result, assignment casts are always planned to convert each insert
column to its target column type, even if the cast ends up being a no-op
because the types are identical.

I attempted to implement a solution that did not plan assignment casts
in all cases, but it proved to be tedious and error prone, so I opted
for this simpler solution for now. My attempt required annotating each
placeholder expression in the memo with a context: a non-assignment
context or an assignment context. This context would be used during
the EXECUTE to determine whether a placeholder value should undergo a
regular cast or an assignment cast: a critical decision due to their
slightly different behaviors (see `PerformAssignmentCast`). The
problem is that correctly annotating placeholders in complex mutations
with expressions like subqueries or CTEs is difficult.

Fixes #71576

There is no release note because this bug was never present in a
release.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72793)
<!-- Reviewable:end -->

rharding6373 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 1 files at r1, 4 of 42 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-:-Moa6gpf5gE_Ae_xgcIy:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @nehageorge, @RaduBerinde, and @rharding6373)

___
*[-- commits, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Moa5OD58Yr_m7uq3WLQ):*
The wording here is ambiguous as to whether the optimization to not plan all assignment casts is implemented in this PR or not, could you please clarify?
___
*[pkg/sql/opt/memo/testdata/logprops/insert, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Moa47fMDVfrCUcpC3MD:-Moa47fMDVfrCUcpC3ME:bcwz8q1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4d6db7d26d70d095df31f98214721ae284e5ae/pkg/sql/opt/memo/testdata/logprops/insert#L26)):*
> ```
>  ├── columns: <none>
>  ├── insert-mapping:
>  │    ├── y:14 => a:1
> ```

Could you explain why the column IDs changed throughout these tests?
___
*[pkg/sql/sem/tree/cast.go, line 1374 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Moa0flV7ECam5xwCTNK:-Moa0flW5iEYalMw7AjI:blnn1g3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4d6db7d26d70d095df31f98214721ae284e5ae/pkg/sql/sem/tree/cast.go#L1374)):*
> ```Go
> // the specified type. The original datum is returned if its type is identical
> // to the specified type. It is similar to PerformCast, but differs because it
> // errors in more cases than PerformCast if the datum's width is too wide for
> ```

nit: enumerate on "more cases"


<!-- Sent from Reviewable.io -->

nehageorge commented on pkg/sql/opt/optbuilder/mutation_builder.go:
```diff
@@ -1393,47 +1393,28 @@ func checkDatumTypeFitsColumnType(col *cat.Column, typ *types.T) {
-			// The column is already the correct type, so add it as a
-			// passthrough column.
-			projectionScope.appendColumn(&inScope.cols[i])
+		// Check if an assignment cast is available from the inScope column
```
Does it make sense to add a comment explaining why we are adding assignment casts in all cases, especially in case we change this in the future?

nehageorge commented on pkg/sql/opt/memo/typing.go:
```diff
@@ -19,9 +19,10 @@ import (
-// the expression's Type field. Depending upon the operator, the type may be
-// fixed, or it may be dependent upon the expression children.
+// InferType derives the type of the given scalar expression. The result is
+// stored it in the expression's Type field by Memoize[Expr] functions.
```
nit: stored it in -> stored in

nehageorge reviewed (commented):
Wow, nice work!

mgartner commented:
There's some things I still need to iron out with this approach... For example, user-defined ENUM types are broken by this change. I'll address the comments already left, but no need to review further until I get the build passing.

RaduBerinde reviewed (commented):
Nice work, looks good. Sorry to send you down a bad path with the placeholder canonical type thing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-:-MoegOpHCGJF0XKLjWG-:b-fluobo)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, and @rharding6373)



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Sorry to send you down a bad path with the placeholder canonical type thing.

It was fun while it lasted.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-:-MoelEq89RxdzTIAA9Lc:b4yf68r)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nehageorge, @RaduBerinde, and @rharding6373)

___
*[-- commits, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Moa5OD58Yr_m7uq3WLQ):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

The wording here is ambiguous as to whether the optimization to not plan all assignment casts is implemented in this PR or not, could you please clarify?
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/typing.go, line 23 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mo_vauZAuZVfYiPqEDk-r2-23:-Moegh0Z5hOXP0cZ2Mt9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4d6db7d26d70d095df31f98214721ae284e5ae/pkg/sql/opt/memo/typing.go#L23)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

nit: stored it in -> stored in
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/testdata/logprops/insert, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Moa47fMDVfrCUcpC3MD:-Moegntg5P_SlftMilK8:b-qem4yi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4d6db7d26d70d095df31f98214721ae284e5ae/pkg/sql/opt/memo/testdata/logprops/insert#L26)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you explain why the column IDs changed throughout these tests?
</blockquote></details>

In this case, column `y:10` contained the insert values for column `abcde.a`, but now that it is assignment-casted before inserting, `y:14` contains the insert values. You can see the projection of `y:14` as the result of `assignment_cast(y:10, INT8)` below in the tree.

In other cases, IDs may have changed simply because we are creating more columns than we did before (columns that are the result of an assignment cast projection). Any column created after the assignment casts are created in optbuilder will have a higher ID than they did before.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1401 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mo_vauZAuZVfYiPqEDp-r2-1401:-MoekBCY0mFfkIi8znZP:b-drxha7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4d6db7d26d70d095df31f98214721ae284e5ae/pkg/sql/opt/optbuilder/mutation_builder.go#L1401)):*
<details><summary><i>Previously, nehageorge wrote…</i></summary><blockquote>

Does it make sense to add a comment explaining why we are adding assignment casts in all cases, especially in case we change this in the future?
</blockquote></details>

Good point. I updated the description for `addAssignmentCasts` above to explain this.
___
*[pkg/sql/sem/tree/cast.go, line 1374 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Moa0flV7ECam5xwCTNK:-MoelCmP4pGJU7RGxdR2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e4d6db7d26d70d095df31f98214721ae284e5ae/pkg/sql/sem/tree/cast.go#L1374)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: enumerate on "more cases"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: add assignment cast normalization rules

This commit adds three new normalization rules related to assignment
casts. These rules work together to eliminate and fold assignment casts
and projections when possible.

These rules are important for allowing fast-path inserts to be planned
in more cases. Fast-path inserts can only be planned when the insert
input is a Values expression, so eliminating assignment cast projections
in mutation inputs is required for this optimization.

These rules also help eliminate uniqueness checks for
`gen_random_uuid()` values in more cases. By pushing down and
eliminating assignment casts that wrap `gen_random_uuid()`, the
optimizer will not plan uniqueness checks in some cases because it can
recognize the builtin function expression.

Release note: None

mgartner committed:
opt: add assign-placeholders-build opt test directive

This commit adds the `assign-placeholders-build` opt test directive. It
behaves similar to `assign-placeholders-norm`, but disables
normalization rules.

Release note: None

mgartner commented:
I've gotten all the kinks ironed out and this is now ready for review. If the Reviewable diff is too crazy from all my rebases and pushes, let me know and I could open a fresh PR.

RaduBerinde reviewed (commented):
Looks great, nice job! Just have one question around the `get_random_uuid()` rule.

> Reviewed 21 of 54 files at r8, 2 of 22 files at r9, 1 of 6 files at r10, 1 of 39 files at r12, 32 of 42 files at r13, 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-:-Mq5vlhN7yoAry9L_wWx:b-u0sfyk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, @RaduBerinde, and @rharding6373)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5uRDf40yH_2lxE4FK:-Mq5uRDf40yH_2lxE4FL:b410ksa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/norm/rules/scalar.opt#L71)):*
> ```
> $input
> 
> # EliminateGenRandomUUIDAssignmentCast discards the assignment cast operator
> ```

This rule seems oddly specific. If a UUID-to-UUID cast is always a no-op (which the code confirms), why can't we remove any such cast, regardless of input expression? (or better yet, never create one in the first place)
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1393 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5ra6f1qec8ovAI4Uw:-Mq5ra6f1qec8ovAI4Ux:b59bs9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e34a862bf93242854dd59b17eaab5d75403a646/pkg/sql/opt/optbuilder/mutation_builder.go#L1393)):*
> ```Go
> // function will error.
> //
> // Assignment casts are always wrap mutation columns even if the type of a
> ```

[nit] remove "are"
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1396 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5rnO09ciw4ZTNH7jb:-Mq5rnO09ciw4ZTNH7jc:b8egddu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e34a862bf93242854dd59b17eaab5d75403a646/pkg/sql/opt/optbuilder/mutation_builder.go#L1396)):*
> ```Go
> // mutation column is identical to its target type. This is required in order to
> // correctly cast placeholder values. When a memo is built at PREPARE-time, the
> // type of a placeholder value provided at EXECUTE-time cannot be known. If a
> ```

[nit] maybe say "the exact type (e.g. int2 vs int8) of a placeholder"
___
*[pkg/sql/opt/testutils/opttester/opt_tester.go, line 1109 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5vWAg1yxkZLSAX2FP:-Mq5vWAg1yxkZLSAX2FQ:b84ns7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d4d474d556c94ee8b3f05274123304948d216ae/pkg/sql/opt/testutils/opttester/opt_tester.go#L1109)):*
> ```Go
> 	queryArgs []string, normalize, explore bool,
> ) (opt.Expr, error) {
> 	disableOptimizations := func(ruleName opt.RuleName) bool {
> ```

[nit] `maybeDisableRule`


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (approved):
I only have a few nits. Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-:-MqFuMsjAzvth84eT6lA:b-in1oyo)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, @RaduBerinde, and @rharding6373)

___
*[pkg/sql/opt/exec/execbuilder/testdata/explain_analyze_plans, line 417 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq1c9KjB0hbFRp8Tdpg:-Mq1c9KjB0hbFRp8Tdph:b4vwi7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze_plans#L417)):*
> ```
> ·
> Diagram 1 (subquery): https://cockroachdb.github.io/distsqlplan/decode.html#eJysUtFq20oQfb9fMcyTDRssOVwo-5Q0uGDiyMV2DKWYsF4NyhJpV90d1XaNP6s_0C8rlqw2IsRtaR_nzJzZM2fPHsOnHCWOk_lotoBxspiCfjR5Csvryf1oDr1YQG8-moxuFlAY2yv78G42vYNSebLc76NA61JKVEEB5UeMUeD_uBJYeqcpBOeP8L4eGqdblJFAY8uKj_BKoHaeUO6RDeeEEhN34crBEAWmxMrk9VLakq7YOAtsCpIQffsaUOBasX6kAK7ismIJEQr0bvMTiHF1ENhUp_cCq4xQXh7EM03xeU0Ltc5pRiolP4i6yhobrkpvCuV3KHBeKhskXKDAiSkMw9GQ22VX-O0StLNM9uVNt0uoT_CkUtmS1zumFnoDb1Hg3fLmBgJTCdpVlqFHWx4Yy30JUa2xGSB6em2gUFsoqHB-ByrPnVZMqYSo3v4XxsZ_Yux1lnnKFDs_iLu-XicfHpLp4iG5n0x6V_ExZv8-BMOO1l8Ec0ahdDZQR-drm6PDSiClGTXhD67ymt57p-tnmnJa82ogpcBN97IpxrZpHQU-J8dnycPz5OFZctQl16fUV6El3jj_BLlisnr3w_kW3yjD3T9JKZA3Kjdf1MsPa2mnmGsyn-kU9bbV5r3tNZlvuwWFoLLOQPS7QVgd_vseAAD___Xaogg=
> Diagram 2 (main-query): https://cockroachdb.github.io/distsqlplan/decode.html#eJyskcFq8zAQhO__Uyx7SkAQK0ed8lNcMKRxSdJeig-uvCQCRXKlVRsIfqy-QJ-s2C60KbQQ6HFmZ0Yf6ITxyaLCYrXJ11soVtsS9N7YBu7_L-_yDUykgMkmX-ZXWzgYN2mncL0ub6CtAzmeTlGg8w2t6gNFVA8osRLYBq8pRh966zQEiuaIKhNoXJu4tyuB2gdCdUI2bAkVWq9rC8-1TRQhm2UosCGujR2Wy8QKFnOxkCiQjqQTG--AzYEUZG-vEQU-1qz3FMEnbvt4PxH8y6chseoEjuqDInK9I1SyE5eSap8cg5zJc9C_Z5tfwram2HoX6Qzqp-WsqwRSs6Pxp6JPQdNt8Hp4ZpTl0BuMhiKPVzmKwo2nHvBrWf5ann8rV92_9wAAAP__OpjUsw==
> ```

After folding the assignment cast, why does this diagram still differ from the base diagram?
___
*[pkg/sql/opt/norm/testdata/rules/fold_constants, line 958 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq1pw2o-5UMRIPOkJhx:-Mq1pw2o-5UMRIPOkJhy:bssynns) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/norm/testdata/rules/fold_constants#L958)):*
> ```
> 
> # Do not fold a stable assignment cast if stable folds are disabled.
> norm no-stable-folds
> ```

can expect-not=FoldAssignmentCast be added here too?
___
*[pkg/sql/sem/tree/constant.go, line 573 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq1b1UL37GZ8g62Tj3y:-Mq1b1UL37GZ8g62Tj3z:b-kdytf2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e34a862bf93242854dd59b17eaab5d75403a646/pkg/sql/sem/tree/constant.go#L573)):*
> ```Go
> 			expr.resString = DString(expr.s)
> 			return NewDNameFromDString(&expr.resString), nil
> ```

nit: remove newline


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nehageorge, @RaduBerinde, and @rharding6373)

___
*[pkg/sql/opt/exec/execbuilder/testdata/explain_analyze_plans, line 417 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq1c9KjB0hbFRp8Tdpg:-MqGFHOdC_qGtIB2AVo9:bypzz9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/exec/execbuilder/testdata/explain_analyze_plans#L417)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

After folding the assignment cast, why does this diagram still differ from the base diagram?
</blockquote></details>

The assignment cast for `SELECT min(p) FROM parent` cannot be folded because its value is not known until exeuction time. You can see in the `EXPLAIN (VERBOSE)` output for this query that the assignment cast is still there:

```
distribution: local
vectorized: true
·
• root
│ columns: ()
│
├── • insert
│   │ columns: ()
│   │ estimated row count: 0 (missing stats)
│   │ into: child(c, p)
│   │
│   └── • buffer
│       │ columns: (column1, column2)
│       │ label: buffer 1
│       │
│       └── • values
│             columns: (column1, column2)
│             size: 2 columns, 1 row
│             row 0, expr 0: crdb_internal.assignment_cast(@S1, NULL::INT8)
│             row 0, expr 1: 1
│
├── • subquery
│   │ id: @S1
│   │ original sql: (SELECT min(p) FROM parent)
│   │ exec mode: one row
│   │
│   └── • group (scalar)
│       │ columns: (any_not_null)
│       │ estimated row count: 1 (missing stats)
│       │ aggregate 0: any_not_null(p)
│       │
│       └── • scan
│             columns: (p)
│             estimated row count: 1 (missing stats)
│             table: parent@parent_pkey
│             spans: LIMITED SCAN
│             limit: 1
│
└── • constraint-check
    │
     ...
```
___
*[pkg/sql/opt/norm/rules/scalar.opt, line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5uRDf40yH_2lxE4FK:-MqG7C6m12z3GfInKFK1:bjhbl95) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/norm/rules/scalar.opt#L71)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This rule seems oddly specific. If a UUID-to-UUID cast is always a no-op (which the code confirms), why can't we remove any such cast, regardless of input expression? (or better yet, never create one in the first place)
</blockquote></details>

You're correct. Because a SQL StrVal is [parsed into a UUID type](https://github.com/cockroachdb/cockroach/blob/d466b09cb34e46cc70834afab44a45ae006df169/pkg/sql/sem/tree/constant.go#L568), and errors if it's not a valid UUID, it should be safe to eliminate the assignment cast for all UUID types. I've made this change. I'll attempt to not create these assignment casts at all in a follow-up PR. It should be valid, but I want to think about it some more.

There's certainly other types that we could eliminate assignment casts for, maybe all except `CHAR(n)`, `VARCHAR(n)`, `"char"`, and `DECIMAL(n, m)`. But I'm inclined to leave it to just UUIDs for now unless we find use-cases for other types. Let me know if you disagree.
___
*[pkg/sql/opt/norm/testdata/rules/fold_constants, line 958 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq1pw2o-5UMRIPOkJhx:-MqGBYD663pwvkEyvBgs:b-mj3rib) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/norm/testdata/rules/fold_constants#L958)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

can expect-not=FoldAssignmentCast be added here too?
</blockquote></details>

Good catch. Added.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1393 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5ra6f1qec8ovAI4Uw:-MqGBfgE6b6LM9oNa_uF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e34a862bf93242854dd59b17eaab5d75403a646/pkg/sql/opt/optbuilder/mutation_builder.go#L1393)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] remove "are"
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 1396 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5rnO09ciw4ZTNH7jb:-MqGBz1tBpJCIB_6PrAj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e34a862bf93242854dd59b17eaab5d75403a646/pkg/sql/opt/optbuilder/mutation_builder.go#L1396)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe say "the exact type (e.g. int2 vs int8) of a placeholder"
</blockquote></details>

Done.
___
*[pkg/sql/opt/testutils/opttester/opt_tester.go, line 1109 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5vWAg1yxkZLSAX2FP:-MqGC8hKC6QaGEyC0E84:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d4d474d556c94ee8b3f05274123304948d216ae/pkg/sql/opt/testutils/opttester/opt_tester.go#L1109)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `maybeDisableRule`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/constant.go, line 573 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq1b1UL37GZ8g62Tj3y:-MqGCDJE-2l29k_WWTJt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e34a862bf93242854dd59b17eaab5d75403a646/pkg/sql/sem/tree/constant.go#L573)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

nit: remove newline
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, @RaduBerinde, and @rharding6373)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5uRDf40yH_2lxE4FK:-MqGSEhd7QxEkiJ2hh1w:b-7qt8hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/norm/rules/scalar.opt#L71)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You're correct. Because a SQL StrVal is [parsed into a UUID type](https://github.com/cockroachdb/cockroach/blob/d466b09cb34e46cc70834afab44a45ae006df169/pkg/sql/sem/tree/constant.go#L568), and errors if it's not a valid UUID, it should be safe to eliminate the assignment cast for all UUID types. I've made this change. I'll attempt to not create these assignment casts at all in a follow-up PR. It should be valid, but I want to think about it some more.

There's certainly other types that we could eliminate assignment casts for, maybe all except `CHAR(n)`, `VARCHAR(n)`, `"char"`, and `DECIMAL(n, m)`. But I'm inclined to leave it to just UUIDs for now unless we find use-cases for other types. Let me know if you disagree.
</blockquote></details>

It's fine to leave it UUID only, but I'd make the rule itself more generic. E.g.

```
# EliminateAsignmentCast discards the assignment cast operator if the input
# type and the target type are identical, and based on the type we know that
# such a cast is always a no-op.
[EliminateAssignmentCast, Normalize]
(AssignmentCast
    $input:*
    $targetTyp:* &
        (HasColType $input $targetTyp) &
        (AssignmentCastIsNoop $targetTyp)
)
=>
$input
```

For now `AssignmentCastIsNoop` can return true only for `UUID` family (and can have a TODO to add more later). This function should also explain the subtility that placeholders can have values that don't conform to their type, which is why we can't elide the cast unconditionally. 


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nehageorge, @RaduBerinde, and @rharding6373)

___
*[pkg/sql/opt/norm/rules/scalar.opt, line 71 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-Mq5uRDf40yH_2lxE4FK:-MqH3GVaB_dFnLuEMfKi:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1a9d8d891f63972ce80d2af9149a1cbd1833747/pkg/sql/opt/norm/rules/scalar.opt#L71)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's fine to leave it UUID only, but I'd make the rule itself more generic. E.g.

```
# EliminateAsignmentCast discards the assignment cast operator if the input
# type and the target type are identical, and based on the type we know that
# such a cast is always a no-op.
[EliminateAssignmentCast, Normalize]
(AssignmentCast
    $input:*
    $targetTyp:* &
        (HasColType $input $targetTyp) &
        (AssignmentCastIsNoop $targetTyp)
)
=>
$input
```

For now `AssignmentCastIsNoop` can return true only for `UUID` family (and can have a TODO to add more later). This function should also explain the subtility that placeholders can have values that don't conform to their type, which is why we can't elide the cast unconditionally. 
</blockquote></details>

Good idea. Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Well done!

> Reviewed 2 of 45 files at r15, 31 of 43 files at r16, 3 of 5 files at r18, 2 of 2 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-:-MqPx4AG53oXWVX6mEtT:b-wehp9g)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @nehageorge, @RaduBerinde, and @rharding6373)

___
*[pkg/sql/opt/norm/general_funcs.go, line 77 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/72793#-MqPwvGg8SjxoMDrZARW:-MqPwvGg8SjxoMDrZARX:b-cc0t8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/230331135bfd7fd8abce31a2edc2d2c7e57ab8c5/pkg/sql/opt/norm/general_funcs.go#L77)):*
> ```Go
> }
> 
> // UUIDType returns the UUID SQL type.
> ```

[nit] we can remove this if it's not used anymore


<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
